### ISSUE TITLE
"illegal_argument_exception" for field starting or ending with a [.]

### ISSUE THREAD

**Issue by danlevy1** [elastic/elasticsearch#28890]
Elasticsearch version: 6.2.1

Plugins installed: None

JVM version: 1.8.0_131

OS version: MacOS 10.13

An `illegal_argument_exception` is thrown with a reason `object field starting or ending with a [.] makes object resolution ambiguous` in certain cases, but not others.

For example, below is an example and the error it throws. Interestingly, because the first and second field names are the same (other than the [.]), the error reason returns `foo..foo`.

```
curl -XPUT 'localhost:9200/test/test/1?pretty' -H 'Content-Type: application/json' -d' {
"foo." : {
"foo" : {
"bar" : 123 }
} }
'
```

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "illegal_argument_exception",
        "reason" : "object field starting or ending with a [.] makes object resolution ambiguous: ['foo..foo']"
      }
    ],
    "type" : "illegal_argument_exception",
    "reason" : "object field starting or ending with a [.] makes object resolution ambiguous: ['foo..foo']"
  },
  "status" : 400
}
```

If I try this example, with a [.] at the end, there is no error.

```
curl -XPUT 'localhost:9200/test/test/1?pretty' -H 'Content-Type: application/json' -d' {
"fooOne." : {
"foo" : {
"bar" : 123 }
} }
'
```

```
{
  "_index" : "test",
  "_type" : "test",
  "_id" : "1",
  "_version" : 7,
  "result" : "updated",
  "_shards" : {
    "total" : 2,
    "successful" : 1,
    "failed" : 0
  },
  "_seq_no" : 6,
  "_primary_term" : 1
}
```

As is expected based on the error reason, the following example gives the correct error.

```
".bar" : {
"foo" : {
"bar" : 123 }
} }
'
```

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "mapper_parsing_exception",
        "reason" : "failed to parse"
      }
    ],
    "type" : "mapper_parsing_exception",
    "reason" : "failed to parse",
    "caused_by" : {
      "type" : "illegal_argument_exception",
      "reason" : "object field starting or ending with a [.] makes object resolution ambiguous: ['.bar']"
    }
  },
  "status" : 400
}
```

Basically, the error reason given is not always accurate.

**Comment by danlevy1** [comment]
I was working on fixing this bug when I stumbled across the above bug. I was wondering how the developers wanted to fix the bug.

Should all fields that end with a [.] be forbidden? In the second example above, a field ending with a [.] succeeds, but the first example fails.

**Comment by dakrone** [comment]
I'm not sure I see the issue in this case, in the first and last examples (`foo.` and `.bar`), there is indeed ambiguity about what field you would mean when you use `foo.<something>` or `<something>.bar`, which is why those are rejected. In the middle example (`fooOne.`), there is no ambiguity with the `fooOne.` field since there are no other fields with the same name, hence there is no error when using that field.
